### PR TITLE
Update sparse file tracker `complete` pointer on progress

### PR DESCRIPTION
**Pull request by tlrx**
This change updates the `complete` pointer when a `Gap` makes progress.

### CODE REVIEW

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java
Comment (henningandersen) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java:35]: Can we add this to the constructor javadoc too. And also mention that this is called prior to notifying listeners (since we'll rely on that), though with out of order possibilities for concurrent notifications.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java
Comment (henningandersen) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java:209]: I'd probably prefer to log a warning instead rather than rethrow. There seems to be no reason to throw out the exception if the consumer were to fail (I am sure it will not for now).

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java
Comment (henningandersen) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java:476]: Should we also change this (can be a separate PR if you prefer) to trigger when `gapRange.start <= complete && gapRange.end > complete`? Seems it does not support fetch from indexing tier well today.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java
Comment (henningandersen) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java:483]: When called from `maybeUpdateCompletePointer`, this assertion depends on `ProgressListenableActionFuture` not notifying of progress on the last byte. It feels slightly subtle to me, but works ofc. The check I suggested in `maybeUpdateCompletePointer` should fix it though.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java
Comment (henningandersen) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java:476]: There is possibly more work to do, like supporting out of order range fetching and still updating the complete pointer correctly. That is definitely a separate PR.

**Review by henningandersen — COMMENTED** [review]
Looks great. Left a few smaller comments, but I think moving forward makes sense.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java
Comment (henningandersen) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java:476]: Disregard this, the merging of ranges done in `onGapSuccess` means this works as is.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java
Comment (tlrx) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java:163]: Relates https://github.com/elastic/elasticsearch-serverless/pull/1872#pullrequestreview-2033093453

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java]: Sure, I pushed b2249a64940c5e19519002d0349f0a6b6da95513. I don't expect this to throw though.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java]: I pushed b15d35150e3dd0d17a59daf908524c1c4ccffd76

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java
Comment (ywangd) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java:163]: Hmm the actual length of the data is `actionResult()` which should be <= `end`. When `actionResult() < end`, I think there should be no listener waiting at all when we reach here? Otherwise these listeners are waiting for a range that can never be fulfilled?

**Review by ywangd — COMMENTED** [review]
With these changes, I wonder whether we still need the shortcircuting for when `progressValue == end` introduced in #108095?

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java:163]: > the actual length of the data is `actionResult()` which should be <= `end`.

I think that the cache does not allow to partially complete ranges, so it expects the range (gap) to be fully written or it considers it failed and any remaining listener will be failed too.

I pushed cdabee72d52b3c814c72f71cbd4296c5b2e329db to ensure it is always completed with a value equals to `end`.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java:127]: The motivation is mostly to reduce contention when updating the `complete` pointer (we have to synchronize on `ranges` on the whole region). 

The contention would be between new ranges to be fetched/completed within the same region (range sizes are typically 128kb, 16mb or 32mb sizes but can also be 1kb or 64kb in searchable snapshots) and the progress made on the first pending range of that region (the progress being updated after every buffer write and we use 2mb buffers or smaller if the range is smaller than 2mb).

I don't think this is a big issue but I don't have a strong opinionnor numbers to decide so I kept a conservative approach.

**Comment by tlrx** [comment]
Thanks for your feedback @ywangd!

> I wonder whether we still need the shortcircuting for when `progressValue == end` introduced in #108095?

I see what you mean bu I still think it makes sense to not complete "early" the listeners due to progress being made if those listeners are waiting for the full gap to be completed. Feels more correct to me.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java
Comment (ywangd) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java:127]: Thanks for explaining it. This reminded me that you and Henning had similar discussion over [the other PR](https://github.com/elastic/elasticsearch/pull/108095#discussion_r1621840567). Sorry for being forgetful. Please disregard my earlier comment.

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java
Comment (ywangd) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java:45]: I think we still need `volatile` here or alternatively make `toString` synchronized?

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java
Comment (ywangd) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java:270]: Mostly making sure I understand how this works. We don't update `complete` unless every byte before it is fetched. We also [adjust](https://github.com/elastic/elasticsearch/blob/7916e6a231db48e73b7cb31889ba25d814c49d6d/x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java#L258) `targetRange.start` so that we can expect one of the ranges will resume fetching from `complete`. And that range is the only one that should update the `complete` field?

**Review by ywangd — COMMENTED** [review]
Thanks for answering my question, Tanguy. I have a few more minor ones. This is looking great. I just need another pass for the tests before approval. Thank you!

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/SparseFileTracker.java:270]: Exactly.

File: x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java
Comment (ywangd) [inline x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java:246]: Nit: extra space :)
```suggestion
            p -> assertThat("LongConsumer should not consumed the same value twice", consumed.add(p), equalTo(true))
```

File: x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java
Comment (ywangd) [inline x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java:239]: The code does not prevent negative `start` values. But it seems rather odd to allow it. I wonder whether we should tighten it up in a separate PR to assert `start >= 0`?

File: x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java
Comment (ywangd) [inline x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java:250]: Nit: Maybe a comment about why `25` is used could be helpful.

File: x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java
Comment (ywangd) [inline x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java:270]: The consumer is not called for `future.onProgress(end)`. Is there any value to assert that? Or do you think this is an implementation detail and not worth it?

File: x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/SparseFileTrackerTests.java
Comment (ywangd) [inline x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/SparseFileTrackerTests.java:565]: I think this branch meant to assert `awaitingKListener.isDone`?

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java
Comment (ywangd) [inline x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java:41]: Can we call out that the consumer is not called when `progress == end`? I find it important and a bit counter-intuitive. So it's helpful to explicitly note it down.

**Review by ywangd — APPROVED** [review]
LGTM

File: x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java:246]: Eagle eyes again :)

File: x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java]: This is just a number of iterations, I used a local variable in 16052d4fa5cf003d0243518f41c08c51a8002314

File: x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/ProgressListenableActionFutureTests.java:271]: I added a check that the consumer is not called when progress is updated to `end` in 16052d4fa5cf003d0243518f41c08c51a8002314

File: x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/SparseFileTrackerTests.java
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/common/SparseFileTrackerTests.java]: Sure, pushed in 16052d4fa5cf003d0243518f41c08c51a8002314

File: x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java
Comment (tlrx) [reply x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java]: Done in 16052d4fa5cf003d0243518f41c08c51a8002314

**Comment by tlrx** [comment]
Thanks Yang & Henning!

**Pull request merged**
